src/parser/ts/index.ts

```diff
@@ -0,0 +1,528 @@
+) => {
+  const tsIds = getReferencedIdsInExpressions(expressions);
+  const tsUniqueIds = uniqBy(tsIds, unwrapId);
+  const referencedTS = await getTimeseries({ items: tsUniqueIds }, target, connector, false);
+  const tsMap = reduceTsToMap(referencedTS);
+  return expressions.map(expr => convertExpressionToLabel(expr, labelSrc, tsMap));
+};
+
+export const getReferencedTimeseries = (
+  route: STSQueryItem[] | STSFunction
+): StringMap[] => {
+  let idFilters: STSServerFilter[] = [];
+  walk(route, obj => {
+    if (isSTSReference(obj)) {
+      const ids = getIdFilters(obj);
+      idFilters = [...idFilters, ...ids];
+    }
+  });
+  return idFilters.map(({ path, value }) => ({
+    [path]: value as string,
+  }));
+};
+
+export const hasAggregates = (
+  expression: string
+): boolean => {
+  let hasAggregates = false;
+  walk(parse(expression), obj => {
+    if (isSTSReference(obj) && obj.query.some(isSTSAggregateFilter)) {
+      hasAggregates = true;
+    }
+  });
+  return hasAggregates
+};
+
+export function flattenServerQueryFilters(items: STSFilter[]): StringMap {
+  return items.reduce((res, filter) => {
+    let value: any;
+    if (isByIdsQuery(filter.value)) {
+      value = filter.value.map(item => flattenServerQueryFilters([item]));
+    } else if(isSTSFilterArr(filter.value)) {
+      value = {...res[filter.path], ...flattenServerQueryFilters(filter.value)};
+    } else if (isArray(filter.value)){
+      value = filter.value.map(flattenServerQueryFilters);
+    } else {
+      value = filter.value
+    }
+    res[filter.path] = value;
+    return res;
+  }, {});
+}
+
+export const walk = (
+  route: STSItem,
+  iterator: (item: STSQueryItem | STSFunction) => any
+) => {
+  if (isArray(route)) {
+    return route.map(r => walk(r, iterator));
+  }
+  const resArr = [];
+  const res = iterator(route);
+  if (res !== undefined) {
+    resArr.push(res);
+  }
+  if (isSTSFunction(route)) {
+    let args: STSItem;
+    if(isMapFunction(route)) {
+      args = route.args[0];
+    } else {
+      args = route.args;
+    }
+    resArr.push(...walk(args, iterator));
+  }
+  return resArr;
+};
+
+export const getIndicesOfMultiaryFunctionArgs = (route: STSQueryItem[] | STSFunction): number[] => {
+  const responseArr = [];
+  const argsIndices = [];
+  let index = 0;
+  walk(route, obj => {
+    if (isMultiaryFunction(obj) && obj.args.length === 1) {
+      argsIndices.push(...obj.args);
+    } else if (isSTSReference(obj) && !hasIdsFilter(obj)) {
+      if (argsIndices.includes(obj)) {
+        responseArr.push(index);
+      }
+      index++;
+    }
+  });
+  return responseArr;
+};
+
+export const getServerFilters = (
+  route: STSQueryItem[] | STSFunction
+): StringMap[] => {
+  const responseArr: StringMap[] = [];
+  const filter = (_, __, parent) => isServerFilter(parent);
+  walk(route, obj => {
+    if (isSTSReference(obj) && !hasIdsFilter(obj)) {
+      const serverQuery = filterDeep(filterDeep(obj.query, filter), filter);
```

Comment:
this can be replaced just with `filterDeep(obj.query, filter)`

Reply:
no, I have test for it, maybe change filter somehow?

Reply:
decided to keep as it is